### DISCUSSION TITLE
It is better to set the capacity of the list.

### DISCUSSION THREAD

**Discussion by behroozbc** [dotnet/MQTTnet#1401] [Ideas]
I just saw this code and noticed a small performance issue.
https://github.com/dotnet/MQTTnet/blob/73d681bc1f978c4e6cf03266fcd1fb4a30a5d205/Source/MQTTnet/Server/Internal/MqttClientSessionsManager.cs#L254

If a list has a small number of items, it does not really matter to set the capacity of the list. However, when the list becomes large, it is important to set the capacity of the list. the cost of performance is `O(log(n))` and happened because of the `EnsureCapacity` method of `List<T>`  
There's a possibility that things might change like this
```
 var result = new List<MqttClientStatus>(_clients.Values.Count);
```
It would be my pleasure to find and fix the whole project if you would like.

**Comment by SeppPenner** [comment]
Sounds legit for me, what do you think of this @chkr1011?

**Reply by chkr1011** [reply]
Yes this is indeed a good optimization. I also applied it at several places but not yet everywhere.

@behroozbc When adding this please make sure to NOT replace the data types. List do support this but HashSet etc. may not.

**Reply by behroozbc** [reply]
Certainly, I'll begin to work on this in small chunks.
